### ISSUE TITLE
How exactly was Figure 5 produced? Reproduction evidence shows a persistent scale gap and unconverged solver outputs

### ISSUE THREAD

**Issue by chenmzh** [microsoft/otdd#36]
## How, exactly, was Figure 5 produced?

This is a reproducibility request supported by a substantially expanded set of experiments, following the unresolved parameter question in #15.

**Please provide the actual configuration and a runnable path from the released code to the Figure 5 numbers.** The public example is not sufficient to reproduce the numerical scale. After testing sampling variability, precision, full-training class statistics, actual full outer supports, annealing, and bounded refinement, we still cannot obtain the reported matrix under the explicitly reconstructed settings below.

In the final NeurIPS 2020 paper, the distance matrix is **Figure 5**, with MNIST–USPS displayed as **1.26 × 10³**. Figure 4 shows label distances and couplings. We recognize that 1.26 × 10³ is rounded; rounding cannot account for the discrepancies reported here.

We are not asking for another suggestion to try unspecified parameters. **Which exact experiment produced this figure, and what settings or transformations are missing from the released reproduction path?**

## 1. Reproducing the public example: 40 attempts, including failures

We used upstream commit `72f1b22d6c688e15db416194c9fe42262d927303` with:

- Python 3.11.16; PyTorch 2.11.0+cu128; torchvision 0.26.0+cu128;
- NumPy 2.4.6; GeomLoss 0.2.6; POT 0.9.7.post1;
- CPU, four threads for this example-reproduction matrix.

This is a **modern compatibility environment, not the original 2020 environment**. Compatibility changes were limited to replacing removed `symeig(..., eigenvectors=True)` calls with `torch.linalg.eigh(..., UPLO='U')`, making the unused torchtext import optional, and converting a cost tensor to NumPy in the POT `emd2` branch. That EMD branch is not used by the default empirical inner Sinkhorn below. These changes did not alter the cost definition, normalization, or covariance approximation.

For each seed 0–9, we independently recreated the loaders and class statistics from official MNIST/USPS data, without reusing a saved OTDD distance/statistics cache. Each method/precision variant used the same selected support.

| Inner method | Precision | Returned a value / attempted | Raw returned cost, mean ± sample SD |
|---|---|---:|---:|
| `exact` — default empirical inner Sinkhorn, **not unregularized EMD** | single | 10/10 | 940.438 ± 12.643 |
| `exact` | double | 10/10 | 940.438 ± 12.643 |
| `gaussian_approx` | single | 0/10 | `torch.linalg.eigh` failures; all retained |
| `gaussian_approx` | double | 10/10 | 880.026 ± 12.970 |

Seed 0, empirical/single: **934.329345703125**. Seeds 0–2 exactly reproduced our earlier archived results. The largest single/double difference for empirical inner Sinkhorn across ten seeds was only **0.0003155**.

A compact reproducer, after applying the compatibility changes and downloading the official data:

```python
import numpy as np
import torch
from otdd.pytorch.datasets import load_torchvision_data
from otdd.pytorch.distance import DatasetDistance

torch.set_num_threads(4)
seed = 0
np.random.seed(seed)
torch.manual_seed(seed)
a = load_torchvision_data(
    'MNIST', resize=28, maxsize=1000,
    datadir='./data', download=False
)[0]['train']
b = load_torchvision_data(
    'USPS', resize=28, maxsize=1000,
    datadir='./data', download=False
)[0]['train']
np.random.seed(seed)
torch.manual_seed(seed)
d = DatasetDistance(
    a, b, inner_ot_method='exact', p=2,
    entreg=.01, inner_ot_entreg=.01,
    device='cpu', precision='single'
)
print(d.distance(maxsamples=10000))
```

The loaders retain the default validation fraction `.1` and select **1,000 training examples per dataset**. `maxsamples=10000` is an upper limit, not a way to turn these loaders into 10,000-example supports. We saved and independently checked the actual indices and class counts. The Gaussian-single failures are observations in this environment, not a claim that the original environment failed identically. Returning a scalar in double precision does not establish Sinkhorn convergence.

## 2. All ten dataset pairs: similar ranking, unresolved numerical scale

Separately from the loader-limited example, we used **fixed full-training Gaussian class statistics**, double precision, outer epsilon `.1`, scaling `.5`, debiasing, and 1,000/2,000 outer samples across **ten sampling seeds**. These are declared reconstructed settings, **not claimed author settings**.

The complete grid contains **1,000 distance records** across five methods, two support sizes, ten dataset pairs, and ten seeds. Before extending the old grid, 50 replay cases matched archived floating-point results exactly.

For the full-Gaussian method at 2,000 outer samples:

| Pair | Our raw returned cost, mean ± sampling SD | Figure 5 displayed value |
|---|---:|---:|
| EMNIST–MNIST | 714.574 ± 2.982 | 1040 |
| FashionMNIST–USPS | 726.292 ± 4.015 | 1100 |
| MNIST–USPS | 856.567 ± 5.803 | 1260 |
| KMNIST–USPS | 874.275 ± 2.983 | 1300 |
| EMNIST–KMNIST | 884.183 ± 2.641 | 1280 |
| EMNIST–USPS | 884.344 ± 3.367 | 1320 |
| KMNIST–MNIST | 960.769 ± 3.447 | 1410 |
| EMNIST–FashionMNIST | 1036.612 ± 7.011 | 1570 |
| FashionMNIST–KMNIST | 1085.739 ± 6.170 | 1670 |
| FashionMNIST–MNIST | 1144.274 ± 7.288 | 1740 |

The agreement in ordering is high: **Spearman = 0.987879** across the ten pairs; EMNIST–MNIST is the closest pair in both matrices. That positive result should not be omitted.

However, the paper's displayed values are **1.448–1.538 times** our measured means. Extending from three to ten sampling seeds changed every Gaussian pair mean by **less than 0.45%**. Ordinary outer-sampling variability therefore does not explain this discrepancy in our tests. A post-hoc common multiplicative fit also does not exactly explain all ten cells, and we have **not rescaled our results** to match the paper.

These default-solver outputs are not presented as convergence-certified optima. That distinction matters for the next set of checks.

## 3. Actual full training supports do not close the gap under this objective

We evaluated **all 60,000 MNIST and 7,291 USPS training examples**, including **both self terms** needed for the debiased quantity—not just the cross problem.

Because the environment lacked PyKeOps and a dense MNIST self-cost matrix alone would require 28.8 GB in float64, we used an independently validated tiled implementation. It evaluates **every pair** with no minibatch, sparse, or low-rank approximation. It uses the fixed full-training Gaussian statistics and cost

`C = (squared_feature_distance + Gaussian_W2_squared) / 2`.

The implementation reproduces the released tensorized callback's cost and annealing schedule on small supports, but it is **not the original backend**. Real n=2000 diagnostics were independently reconstructed with dense NumPy calculations.

At epsilon `.1`:

| Full supports: 60000 / 7291 | Raw debiased potential scalar | Numerical debiased lower / upper bounds | Cross row / column L1 residual |
|---|---:|---:|---:|
| scaling `.5` | 843.279047 | [843.011895, 978.630862] | 2.692873 / 3.069457 |
| scaling `.9` | 858.823281 | [858.824956, 897.606059] | 0.147628 / 0.150744 |

**Neither full-support solve converged.** The bounds use mass-corrected dual objectives and numerically feasible rounded-plan primal objectives, with the self bounds combined using the appropriate signs. They are floating-point numerical bounds, **not directed-rounding interval-arithmetic certificates**. Full-support saved vectors and arithmetic were independently checked; we did not independently reconstruct a second full-support cost/plan implementation.

The raw potential scalar can lie slightly below the corrected lower bound: away from feasibility it is not itself a feasible primal objective or the corrected dual. We retained that behavior instead of clipping it away or calling the scalar an optimum.

**Under this fixed reconstructed objective, increasing the outer support to the entire training sets does not bridge the gap to 1260.** The `.9` numerical upper bound is approximately 897.606. This does not establish what objective the authors actually used; it makes the missing protocol details more important, not less.

## 4. A plausible-looking returned scalar is not evidence of convergence

The bounded scale/solver study contained **36 invocations**, including the two full-support runs and 34 subset/annealing/epsilon controls. All completed without process failures or timeouts. We retained **179 diagnostic checkpoints and their saved potential-vector archives**.

Across the **40 baseline/final-refined debiased views**, **none passed all predeclared convergence criteria**: both marginal L1 residuals at most `1e-8` and a numerical primal-dual gap at most `1e-4` for each cross/self component, together with repaired-plan feasibility checks. Some individual components converged; the complete debiased quantities did not.

- At n=1000/2000, scaling `.99` improved cross L1 residuals to approximately **0.014–0.018**, still far from the marginal criterion.
- In the separate n=2000, seed-0 epsilon `.01/.1/1/10` sensitivity test, we preserved each `.5` baseline before at most 200 additional alternating fixed-epsilon steps.
- At epsilon `.01`, the baseline returned a superficially plausible raw debiased scalar near **853.496**, while the implied cross-coupling mass was approximately **1.84 × 10¹²**.
- Epsilon `10` with refinement improved the debiased numerical interval to **[854.776370, 854.778495]**, but the worst marginal L1 residual was still about **1.32 × 10⁻⁵**, so we did not relabel it as converged.

Changing epsilon changes the regularized objective. We did not select an epsilon or an output transformation because it happened to approach a paper value. These finite tests also do **not** claim that a sufficiently long or different solver could never converge.

## Our conclusions

1. **We cannot reproduce Figure 5's numerical scale from the released example and the explicitly reconstructed configurations tested here.** This is a reproducibility gap that requires an actual experimental specification, not an undocumented parameter guess.
2. **The discrepancy is not explained by the sampling repetition count or by empirical-inner single versus double precision in these tests.** Actual full outer supports do not close it under our fixed Gaussian objective either.
3. **The distance ranking is substantially reproduced.** We are not claiming that every qualitative result fails.
4. **Default returned scalars must not be treated as converged OT solutions without diagnostics.** Our tests expose substantial marginal/optimality problems, including cases whose returned scalar looks entirely ordinary.
5. **These findings do not establish fabrication or invalidate the paper under an unknown original protocol.** They establish that important settings and the figure-generation path remain unresolved, and that numerical convergence needs to be checked rather than assumed.

## What we need from the maintainers/authors

Please answer these concretely, preferably with the actual figure-generation script, config, and unrounded result table:

1. **Which commit, dependency versions, script, and command produced Figure 5?** Is that exact path present in this repository? If not, can it be released?
2. **What were the class-statistics pools and outer supports?** Give the train/test split, per-dataset counts, validation fraction, sampling scheme/seeds, and whether both stages used the same subset.
3. **What preprocessing and ground cost were used?** In particular: 28×28 USPS resizing, shared normalization `(0.1307, 0.3081)`, EMNIST orientation, full versus approximate covariance, covariance normalization, and the `1/p` factor for `p=2`.
4. **What were the inner and outer solver settings?** Inner method and epsilon, outer epsilon, debiasing for each level, precision, backend, annealing/scaling, stopping rule, and any additional iterations.
5. **What exactly was plotted?** The raw returned cost, a square root, a scaling, a sum/mean across repeats, or another transformation? Please supply the unrounded numbers.
6. **How was convergence checked for the published matrix?** Were marginal residuals or primal-dual gaps inspected? If not, what evidence establishes that the reported values were sufficiently solved?

**If the original scripts or settings are no longer available, please say so explicitly and document that limitation. If the public example was never meant to reproduce the figure, please identify what additional steps are required.**

We have preserved the complete result grids, actual sampling indices, compatibility patch, environment identities, failures, saved solver vectors, and independent audit results. The evidence above is not based on a single unlucky seed or an unexplained screenshot. A concrete, reproducible explanation of how the published numbers were obtained is the appropriate next step.
